pkg/webhook/pod/pod.go

```diff
@@ -18,95 +18,69 @@ package pod
 
 import (
 	"context"
+	"encoding/json"
 	"net/http"
 
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/manager"
-	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
-	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
-	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/builder"
-	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/types"
-
-	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	"github.com/k8snetworkplumbingwg/kubemacpool/pkg/pool-manager"
+	webhookserver "github.com/qinqon/kube-admission-webhook/pkg/webhook/server"
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
-
-	"github.com/k8snetworkplumbingwg/kubemacpool/pkg/pool-manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
-var log = logf.Log.WithName("Webhook Pod")
-
-func Add(mgr manager.Manager, poolManager *pool_manager.PoolManager, namespaceSelector *v1.LabelSelector) (*admission.Webhook, error) {
-	podAnnotator := &podAnnotator{poolManager: poolManager}
-
-	wh, err := builder.NewWebhookBuilder().
-		Mutating().
-		FailurePolicy(admissionregistrationv1beta1.Fail).
-		Operations(admissionregistrationv1beta1.Create).
-		ForType(&corev1.Pod{}).
-		Handlers(podAnnotator).
-		WithManager(mgr).
-		NamespaceSelector(namespaceSelector).
-		Build()
-	if err != nil {
-		return nil, err
-	}
-
-	return wh, nil
-}
+var log = logf.Log.WithName("Webhook mutatepods")
 
 type podAnnotator struct {
 	client      client.Client
-	decoder     types.Decoder
+	decoder     *admission.Decoder
 	poolManager *pool_manager.PoolManager
 }
 
-// podAnnotator Implements admission.Handler.
-var _ admission.Handler = &podAnnotator{}
+// Add adds server modifiers to the server, like registering the hook to the webhook server.
+func Add(s *webhookserver.Server, poolManager *pool_manager.PoolManager) error {
+	podAnnotator := &podAnnotator{poolManager: poolManager}
+	s.UpdateOpts(webhookserver.WithHook("/mutate-pods", &webhook.Admission{Handler: podAnnotator}))
+	return nil
+}
 
 // podAnnotator adds an annotation to every incoming pods.
-func (a *podAnnotator) Handle(ctx context.Context, req types.Request) types.Response {
+func (a *podAnnotator) Handle(ctx context.Context, req admission.Request) admission.Response {
 	pod := &corev1.Pod{}
 
 	err := a.decoder.Decode(req, pod)
 	if err != nil {
-		return admission.ErrorResponse(http.StatusBadRequest, err)
+		return admission.Errored(http.StatusBadRequest, err)
 	}
 
 	if pod.Annotations == nil {
-		admission.PatchResponse(pod, pod)
```

Comment:
This should not be removed. Here we do early return in case there are no annotations (no extra network requested). In your PR you initialize the annotation.

Reply:
I didn't see an early return in the master branch ([permalink](https://github.com/k8snetworkplumbingwg/kubemacpool/blob/599529e12198a7c3e0f7aef5bdb60fc514ad045b/pkg/webhook/pod/pod.go#L78))
Nonetheless I will add one.

Reply:
Thanks!

Reply:
Adding early return fails the tests. currently dropping this change and will revisit it in another commit.